# 🧠 Day 3 – SQL via Python: NYC School Data Exploration
In this notebook, you'll connect to a PostgreSQL database and execute SQL queries to explore NYC school data.

## 🔌 Step 1: Import Libraries

In [114]:
import pandas as pd
import psycopg2

## 🔐 Step 2: Connect to the Database

In [115]:
# DB connection setup using hardcoded credentials (for onboarding only)
conn = psycopg2.connect(
    dbname="neondb",
    user="neondb_owner",
    password="npg_CeS9fJg2azZD",
    host="ep-falling-glitter-a5m0j5gk-pooler.us-east-2.aws.neon.tech",
    port="5432",
    sslmode="require"
)
cur = conn.cursor()

In [116]:
print(cur)

<cursor object at 0x11a105e40; closed: 0>


## 🔍 Step 3: Run queries to load the tables

In [117]:
# Loading table high_school_directory into a pandas DataFrame
query_1 = "SELECT * FROM nyc_schools.high_school_directory;"
pd.set_option('display.max_columns', None)
df_sch_dir = pd.read_sql(query_1, conn)
df_sch_dir.head(5)

/var/folders/hj/bjtkr_b13175_g6c53_yly7c0000gn/T/ipykernel_23386/3747029791.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_sch_dir = pd.read_sql(query_1, conn)


dbn                                        school_name    borough  \
0  27Q260          Frederick Douglass Academy VI High School     Queens   
1  21K559        Life Academy High School for Film and Music   Brooklyn   
2  16K393     Frederick Douglass Academy IV Secondary School   Brooklyn   
3  08X305                               Pablo Neruda Academy      Bronx   
4  03M485  Fiorello H. LaGuardia High School of Music & A...  Manhattan   

  building_code  phone_number    fax_number  grade_span_min  grade_span_max  \
0          Q465  718-471-2154  718-471-2890             9.0              12   
1          K400  718-333-7750  718-333-7775             9.0              12   
2          K026  718-574-2820  718-574-2821             9.0              12   
3          X450  718-824-1682  718-824-1663             9.0              12   
4          M485  212-496-0700  212-724-5748             9.0              12   

   expgrade_span_min  expgrade_span_max start_time end_time  \
0                NaN                NaN    7:45 AM  2:05 PM   
1                NaN                NaN    8:15 AM  3:00 PM   
2                NaN                NaN    8:00 AM  2:20 PM   
3                NaN                NaN    8:00 AM  3:50 PM   
4                NaN                NaN    8:00 AM  4:00 PM   

                                          priority01  \
0  Priority to Queens students or residents who a...   
1  Priority to New York City residents who attend...   
2                 Priority to continuing 8th graders   
3  Priority to Bronx students or residents who at...   
4                    Open to New York City residents   

                                          priority02  \
0  Then to New York City residents who attend an ...   
1                    Then to New York City residents   
2  Then to Brooklyn students or residents who att...   
3  Then to New York City residents who attend an ...   
4  Admission is based on the outcome of a competi...   

                                          priority03  \
0               Then to Queens students or residents   
1                                                      
2  Then to New York City residents who attend an ...   
3                Then to Bronx students or residents   
4  Students must audition for each program (studi...   

                                          priority04  \
0                    Then to New York City residents   
1                                                      
2             Then to Brooklyn students or residents   
3                    Then to New York City residents   
4  Students must be residents of New York City at...   

                        priority05 priority06 priority07 priority08  \
0                                                                     
1                                                                     
2  Then to New York City residents                                    
3                                                                     
4                                                                     

  priority09 priority10 location phone_number2 school_email  \
0                           None          None         None   
1                           None          None         None   
2                           None          None         None   
3                           None          None         None   
4                           None          None         None   

                                        website  \
0  http://schools.nyc.gov/schoolportals/27/Q260   
1  http://schools.nyc.gov/schoolportals/21/K559   
2  http://schools.nyc.gov/schoolportals/16/K393   
3                    www.pablonerudaacademy.org   
4                           www.laguardiahs.org   

                                              subway  \
0                       A to Beach 25th St-Wavecrest   
1  D to 25th Ave ; N to Ave U ; N to Gravesend - ...   
2             J to Kosciusko St ; M, Z to Myrtle Ave   
3               

In [118]:
# Loading table school_demographics into a pandas DataFrame
query_2 = "SELECT * FROM nyc_schools.school_demographics;"
#pd.set_option('display.max_columns', None)
df_sch_demo = pd.read_sql(query_2, conn)
df_sch_demo.head(5)

/var/folders/hj/bjtkr_b13175_g6c53_yly7c0000gn/T/ipykernel_23386/1082616859.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_sch_demo = pd.read_sql(query_2, conn)


dbn                       Name  schoolyear fl_percent  frl_percent  \
0  01M015  P.S. 015 ROBERTO CLEMENTE    20052006       89.4          NaN   
1  01M015  P.S. 015 ROBERTO CLEMENTE    20062007       89.4          NaN   
2  01M015  P.S. 015 ROBERTO CLEMENTE    20072008       89.4          NaN   
3  01M015  P.S. 015 ROBERTO CLEMENTE    20082009       89.4          NaN   
4  01M015  P.S. 015 ROBERTO CLEMENTE    20092010                    96.5   

   total_enrollment prek   k grade1 grade2 grade3 grade4 grade5 grade6 grade7  \
0               281   15  36     40     33     38     52     29     38          
1               243   15  29     39     38     34     42     46                 
2               261   18  43     39     36     38     47     40                 
3               252   17  37     44     32     34     39     49                 
4               208   16  40     28     32     30     24     38                 

  grade8 grade9 grade10 grade11 grade12  ell_num  ell_percent  sped_num  \
0                                           36.0         12.8        57   
1                                           38.0         15.6        55   
2                                           52.0         19.9        60   
3                                           48.0         19.0        62   
4                                           40.0         19.2        46   

   sped_percent  ctt_num  selfcontained_num  asian_num  asian_per  black_num  \
0          20.3     25.0                9.0         10        3.6         74   
1          22.6     19.0               15.0         18        7.4         68   
2          23.0     20.0               14.0         16        6.1         77   
3          24.6     21.0               17.0         16        6.3         75   
4          22.1     14.0               14.0         16        7.7         67   

   black_per  hispanic_num  hispanic_per  white_num  white_per  male_num  \
0       26.3           189          67.3          5        1.8       158   
1       28.0           153          63.0          4        1.6       140   
2       29.5           157          60.2          7        2.7       143   
3       29.8           149          59.1          7        2.8       149   
4       32.2           118          56.7          6        2.9       124   

   male_per  female_num  female_per  
0      56.2         123        43.8  
1      57.6         103        42.4  
2      54.8         118        45.2  
3      59.1         103        40.9  
4      59.6          84        40.4

In [119]:
# Loading table school_safety_report into a pandas DataFrame
query_3 = "SELECT * FROM nyc_schools.school_safety_report;"
#pd.set_option('display.max_columns', None)
df_sch_safety = pd.read_sql(query_3, conn)
df_sch_safety.head(5)

/var/folders/hj/bjtkr_b13175_g6c53_yly7c0000gn/T/ipykernel_23386/689606178.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_sch_safety = pd.read_sql(query_3, conn)


school_year building_code     dbn                 location_name  \
0     2015-16          K006  17K006                      P.S. 006   
1     2015-16          K005  16K005    P.S. 005 Dr. Ronald Mcnair   
2     2015-16          K003  13K003  P.S. 003 The Bedford Village   
3     2015-16          K002  75K141                     P.S. K141   
4     2015-16          K002  84K704    EXPLORE CHARTER SCHOOL(BS)   

  location_code              address borough  geographical_district_code  \
0          K006     43 SNYDER AVENUE       K                        17.0   
1          K005   820 HANCOCK STREET       K                        16.0   
2          K003  50 JEFFERSON AVENUE       K                        13.0   
3          K141  655 PARKSIDE AVENUE       K                        17.0   
4          K704  655 PARKSIDE AVENUE       K                        17.0   

  register                              building_name num_schools  \
0      712                                                   None   
1      231                                                   None   
2      450                                                   None   
3      374  655 PARKSIDE AVENUE CONSOLIDATED LOCATION        None   
4      529  655 PARKSIDE AVENUE CONSOLIDATED LOCATION        None   

                                 schools_in_building  major_n  oth_n  \
0                                           P.S. 006      0.0    1.0   
1                         P.S. 005 Dr. Ronald Mcnair      0.0    0.0   
2                       P.S. 003 The Bedford Village      0.0    0.0   
3  Parkside Preparatory Academy |  P.S. K141 |Exp...      NaN    NaN   
4  Parkside Preparatory Academy |  P.S. K141 |Exp...      NaN    NaN   

   nocrim_n  prop_n  vio_n engroupa   rangea  avgofmajor_n  avgofoth_n  \
0       2.0     1.0    0.0       4C  501-750          0.28        1.11   
1       2.0     0.0    0.0       2C    1-250          0.24        0.71   
2       2.0     0.0    0.0       3C  251-500          0.27        0.90   
3       NaN     NaN    NaN       3C  251-500           NaN         NaN   
4       NaN     NaN    NaN       4C  501-750           NaN         NaN   

   avgofnocrim_n  avgofprop_n  avgofvio_n borough_name  postcode   latitude  \
0           1.74         0.75        0.53    BROOKLYN    11226.0  40.648590   
1           1.71         0.51        0.32    BROOKLYN    11233.0  40.685695   
2           1.67         0.63        0.44    BROOKLYN    11216.0  40.682663   
3            NaN          NaN         NaN    BROOKLYN    11226.0  40.656183   
4            NaN          NaN         NaN    BROOKLYN    11226.0  40.656183   

   longitude  community_board  council_district  census_tract        bin  \
0 -73.957138             14.0              40.0         794.0  3117230.0   
1 -73.922463              3.0              41.0         377.0  3039906.0   
2 -73.955338              3.0              36.0         227.0  3057406.0   
3 -73.951583              9.0              40.0         802.0  3378201.0   
4 -73.951583              9.0              40.0         802.0  3378201.0   

            bbl                                                nta  _schools  
0  3.051030e+09  Erasmus                                       ...         1  
1  3.014900e+09  Stuyvesant Heights                            ...         1  
2  3.019980e+09  Clinton Hill                                  ...         1  
3  3.050500e+09  Prospect Lefferts Gardens-Wingate             ...         3  
4  3.050500e+09  Prospect Lefferts Gardens-Wingate             ...         3

## ✅ Task Queries Below

### **🧮 School Distribution**

How many schools are there in each borough?

In [120]:
# Example: Count schools by borough
query = "SELECT borough, COUNT(*) AS school_count FROM nyc_schools.high_school_directory GROUP BY borough;"

df_result = pd.read_sql(query, conn)
df_result

/var/folders/hj/bjtkr_b13175_g6c53_yly7c0000gn/T/ipykernel_23386/2881711143.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_result = pd.read_sql(query, conn)


borough  school_count
0       Brooklyn           121
1         Queens            80
2  Staten Island            10
3      Manhattan           106
4          Bronx           118

In [121]:
# Example: Distinct Count of schools by borough
query_schoolcount_ = "SELECT borough, COUNT(DISTINCT dbn) AS unique_school_count FROM nyc_schools.high_school_directory GROUP BY borough ORDER BY unique_school_count DESC;"

df_result_ = pd.read_sql(query_schoolcount_, conn)
df_result_

/var/folders/hj/bjtkr_b13175_g6c53_yly7c0000gn/T/ipykernel_23386/323236893.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_result_ = pd.read_sql(query_schoolcount_, conn)


borough  unique_school_count
0       Brooklyn                  121
1          Bronx                  118
2      Manhattan                  106
3         Queens                   80
4  Staten Island                   10

✅ Summary:

Brooklyn has the highest number of unique schools (121), followed closely by the Bronx (118) and Manhattan (106).
Staten Island has the fewest, with only 10 unique schools listed in the dataset. Normal school count and unique school counts are same.

### **🎓 Language Learners**

What is the average % of English Language Learners (ELL) per borough?

In [122]:
# Count average ELL % per borough (excluding unmatched rows)
query_per_ell = """
SELECT hsd.borough, ROUND(AVG(sd.ell_percent)::numeric, 2) AS avg_ell_percentage
FROM nyc_schools.high_school_directory hsd
Left JOIN nyc_schools.school_demographics sd 
    ON  hsd.dbn  = sd.dbn

GROUP BY hsd.borough
ORDER BY avg_ell_percentage DESC;
"""

df_result_ell = pd.read_sql(query_per_ell, conn)
df_result_ell

/var/folders/hj/bjtkr_b13175_g6c53_yly7c0000gn/T/ipykernel_23386/2845486948.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_result_ell = pd.read_sql(query_per_ell, conn)


borough  avg_ell_percentage
0       Brooklyn                 NaN
1         Queens                 NaN
2  Staten Island                 NaN
3          Bronx                 NaN
4      Manhattan                7.57

**🔍 Final Summary:**

- Only Manhattan has available and valid data for English Language Learner (ELL) percentage.
- The average english language learning percentage in Manhattan is 7.57%.
- No other boroughs (Brooklyn, Bronx, Queens, Staten Island) had sufficient ell_percent data to be included in the result.

### **🔗School supporting special needs**

Using the data from the school demographics and high school directory, write a query to find the top 3 schools in each borough with the highest percentage of special education students (sped_percent)

In [129]:
# top 3 school name in each borough with highest percentage of special education students (sped_percent)
query_per_sped_ = """
SELECT *
FROM (
    SELECT 
        hsd.borough,
        hsd.school_name,
        MAX(sd.sped_percent) AS sped_percent,
        ROW_NUMBER() OVER (PARTITION BY hsd.borough ORDER BY MAX(sd.sped_percent) DESC) AS rank
    FROM nyc_schools.school_demographics sd
    JOIN nyc_schools.high_school_directory hsd 
        ON sd.dbn = hsd.dbn
    WHERE sd.sped_percent IS NOT NULL
    GROUP BY hsd.borough, hsd.school_name
) ranked
WHERE rank <= 3;
"""

df_result_sped_ = pd.read_sql(query_per_sped_, conn)
df_result_sped_

/var/folders/hj/bjtkr_b13175_g6c53_yly7c0000gn/T/ipykernel_23386/2332638113.py:19: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_result_sped_ = pd.read_sql(query_per_sped_, conn)


borough                                    school_name  sped_percent  \
0  Manhattan                     East Side Community School          28.8   
1  Manhattan                        Marta Valle High School          25.9   
2  Manhattan  Henry Street School for International Studies          25.1   

   rank  
0     1  
1     2  
2     3

**🟦 Key Insight: Special Education Data by Borough**

- ✅ Manhattan is the only borough in the merged dataset that contains valid (non-null) sped_percent (special - education percentage) data.
